File: games/spring2026/agents/bot.py
#!/usr/bin/env python3
# Spring Challenge 2026 (Troll Farm) - simple heuristic bot.
# Each troll: drop near shack, harvest if standing on a fruited tree, wait on
# a size-4 tree empty-handed, otherwise head for the most valuable tree by
# Manhattan distance.

import sys

PLAYER_ME = 0
PLAYER_OP = 1

ITEM_PLUM = 0
ITEM_LEMON = 1
ITEM_APPLE = 2
ITEM_BANANA = 3
ITEM_IRON = 4
ITEM_WOOD = 5
NUM_ITEMS = 6

FRUIT_NAMES = {'PLUM': ITEM_PLUM, 'LEMON': ITEM_LEMON, 'APPLE': ITEM_APPLE, 'BANANA': ITEM_BANANA}

DIRS = [(0, -1), (1, 0), (0, 1), (-1, 0)]

width, height = map(int, input().split())
cells = [['.'] * width for _ in range(height)]
shack = [-1, -1]

for y in range(height):
    row = input()
    for x in range(width):
        cells[y][x] = row[x]
        if row[x] == '0':
            shack[PLAYER_ME] = y * width + x
        elif row[x] == '1':
            shack[PLAYER_OP] = y * width + x


def adj_shack_me(c):
    x, y = c % width, c // width
    for dx, dy in DIRS:
        nx, ny = x + dx, y + dy
        if 0 <= nx < width and 0 <= ny < height and cells[ny][nx] == '0':
            return True
    return False


def tree_at(trees, cell):
    for t in trees:
        if t['cell'] == cell:
            return t
    return None


def best_tree_target(trees, from_cell):
    fx, fy = from_cell % width, from_cell // width
    best = -1
    best_dist = 0
    best_tier = -1
    for t in trees:
        tx, ty = t['cell'] % width, t['cell'] // width
        dist = abs(fx - tx) + abs(fy - ty)
        if t['fruits'] > 0:
            tier = 2
        elif t['size'] == 4:
            tier = 1
        else:
            tier = 0
        if tier > best_tier or (tier == best_tier and (best < 0 or dist < best_dist)):
            best = t['cell']
            best_dist = dist
            best_tier = tier
    return best


def decide_troll(tr, trees):
    cell = tr['cell']
    carry = sum(tr['carry'])
    fruits = tr['carry'][ITEM_PLUM] + tr['carry'][ITEM_LEMON] + tr['carry'][ITEM_APPLE] + tr['carry'][ITEM_BANANA]
    full = carry >= tr['carryCap']

    if carry > 0 and adj_shack_me(cell):
        return f"DROP {tr['id']}"

    if not full:
        tree = tree_at(trees, cell)
        if tree is not None:
            if tree['fruits'] > 0:
                return f"HARVEST {tr['id']}"
            if fruits == 0 and tree['size'] == 4:
                return "WAIT"

    if full or fruits > 0:
        s = shack[PLAYER_ME]
        return f"MOVE {tr['id']} {s % width} {s // width}"

    target = best_tree_target(trees, cell)
    if target >= 0:
        return f"MOVE {tr['id']} {target % width} {target // width}"
    return "WAIT"


# game loop
while True:
    try:
        input()  # my inventory (unused)
        input()  # opp inventory (unused)
    except EOFError:
        break

    tree_count = int(input())
    trees = []
    for _ in range(tree_count):
        parts = input().split()
        trees.append({
            'cell': int(parts[2]) * width + int(parts[1]),
            'type': FRUIT_NAMES.get(parts[0], ITEM_PLUM),
            'size': int(parts[3]),
            'health': int(parts[4]),
            'fruits': int(parts[5]),
            'cooldown': int(parts[6]),
        })

    troll_count = int(input())
    trolls = []
    for _ in range(troll_count):
        p = input().split()
        trolls.append({
            'id': int(p[0]),
            'player': int(p[1]),
            'cell': int(p[3]) * width + int(p[2]),
            'moveSpeed': int(p[4]),
            'carryCap': int(p[5]),
            'harvestPower': int(p[6]),
            'chopPower': int(p[7]),
            'carry': [int(p[8]), int(p[9]), int(p[10]), int(p[11]), int(p[12]), int(p[13])],
        })

    cmds = [decide_troll(tr, trees) for tr in trolls if tr['player'] == PLAYER_ME]
    print(';'.join(cmds) if cmds else 'WAIT')

File: games/spring2026/agents/bot.cpp
// Spring Challenge 2026 (Troll Farm) - simple heuristic bot.
// Each troll: drop near shack, harvest if standing on a fruited tree, wait on
// a size-4 tree empty-handed, otherwise head for the most valuable tree by
// Manhattan distance.

#include <iostream>
#include <string>
#include <vector>
#include <array>
#include <climits>
#include <cstdlib>

using namespace std;

static const int PLAYER_ME = 0;
static const int PLAYER_OP = 1;

static const int ITEM_PLUM = 0;
static const int ITEM_LEMON = 1;
static const int ITEM_APPLE = 2;
static const int ITEM_BANANA = 3;
static const int ITEM_IRON = 4;
static const int ITEM_WOOD = 5;
static const int NUM_ITEMS = 6;

static const int DX[4] = {0, 1, 0, -1};
static const int DY[4] = {-1, 0, 1, 0};

static int gWidth = 0;
static int gHeight = 0;
static vector<char> gCell;
static vector<bool> gWalkable;
static vector<array<bool, 2>> gAdjShack;
static int gShack[2] = {-1, -1};

static inline int xy(int x, int y) { return y * gWidth + x; }
static inline int cellX(int c) { return c % gWidth; }
static inline int cellY(int c) { return c / gWidth; }
static inline bool inBounds(int x, int y) {
    return x >= 0 && x < gWidth && y >= 0 && y < gHeight;
}

struct Tree {
    int cell;
    int type;
    int size;
    int health;
    int fruits;
    int cooldown;
};

struct Troll {
    int id;
    int player;
    int cell;
    int moveSpeed;
    int carryCap;
    int harvestPower;
    int chopPower;
    int carry[NUM_ITEMS];

    int carryTotal() const {
        int s = 0;
        for (int i = 0; i < NUM_ITEMS; i++) s += carry[i];
        return s;
    }
    int fruitTotal() const {
        return carry[ITEM_PLUM] + carry[ITEM_LEMON] + carry[ITEM_APPLE] + carry[ITEM_BANANA];
    }
};

static int parseFruit(const string& s) {
    if (s == "PLUM") return ITEM_PLUM;
    if (s == "LEMON") return ITEM_LEMON;
    if (s == "APPLE") return ITEM_APPLE;
    if (s == "BANANA") return ITEM_BANANA;
    return ITEM_PLUM;
}

static void readBoard() {
    cin >> gWidth >> gHeight;
    cin.ignore();
    int n = gWidth * gHeight;
    gCell.assign(n, '.');
    gWalkable.assign(n, false);
    gAdjShack.assign(n, {false, false});

    for (int y = 0; y < gHeight; y++) {
        string row;
        getline(cin, row);
        for (int x = 0; x < gWidth; x++) {
            int c = xy(x, y);
            char ch = row[x];
            gCell[c] = ch;
            if (ch == '.') gWalkable[c] = true;
            if (ch == '0') gShack[PLAYER_ME] = c;
            if (ch == '1') gShack[PLAYER_OP] = c;
        }
    }

    for (int c = 0; c < n; c++) {
        int x = cellX(c), y = cellY(c);
        for (int d = 0; d < 4; d++) {
            int nx = x + DX[d], ny = y + DY[d];
            if (!inBounds(nx, ny)) continue;
            int nb = xy(nx, ny);
            if (gCell[nb] == '0') gAdjShack[c][PLAYER_ME] = true;
            if (gCell[nb] == '1') gAdjShack[c][PLAYER_OP] = true;
        }
    }
}

static int absDiff(int a, int b) { return a < b ? b - a : a - b; }

static const Tree* treeAt(const vector<Tree>& trees, int cell) {
    for (const auto& t : trees) {
        if (t.cell == cell) return &t;
    }
    return nullptr;
}

static int bestTreeTarget(const vector<Tree>& trees, int from) {
    int fx = cellX(from), fy = cellY(from);
    int best = -1;
    int bestDist = INT_MAX;
    int bestTier = -1;
    for (const auto& t : trees) {
        int tx = cellX(t.cell), ty = cellY(t.cell);
        int dist = absDiff(fx, tx) + absDiff(fy, ty);
        int tier = 0;
        if (t.fruits > 0) tier = 2;
        else if (t.size == 4) tier = 1;
        if (tier > bestTier || (tier == bestTier && (best < 0 || dist < bestDist))) {
            best = t.cell;
            bestDist = dist;
            bestTier = tier;
        }
    }
    return best;
}

static string decideTroll(const Troll& tr, const vector<Tree>& trees) {
    int cell = tr.cell;
    int carry = tr.carryTotal();
    int fruits = tr.fruitTotal();
    bool full = carry >= tr.carryCap;

    if (carry > 0 && gAdjShack[cell][PLAYER_ME]) {
        return "DROP " + to_string(tr.id);
    }

    if (!full) {
        if (const Tree* tree = treeAt(trees, cell)) {
            if (tree->fruits > 0) return "HARVEST " + to_string(tr.id);
            if (fruits == 0 && tree->size == 4) return "WAIT";
        }
    }

    if (full || fruits > 0) {
        int s = gShack[PLAYER_ME];
        return "MOVE " + to_string(tr.id) + " " + to_string(cellX(s)) + " " + to_string(cellY(s));
    }

    int target = bestTreeTarget(trees, cell);
    if (target >= 0) {
        return "MOVE " + to_string(tr.id) + " " + to_string(cellX(target)) + " " + to_string(cellY(target));
    }
    return "WAIT";
}

int main() {
    ios::sync_with_stdio(false);

    readBoard();

    while (true) {
        int p, l, a, b, ir, wd;
        if (!(cin >> p >> l >> a >> b >> ir >> wd)) break;
        // opponent inventory (unused)
        int op[6];
        for (int i = 0; i < 6; i++) cin >> op[i];

        int treeCount;
        cin >> treeCount;
        vector<Tree> trees;
        trees.reserve(treeCount);
        for (int i = 0; i < treeCount; i++) {
            string type;
            int x, y, size, health, fruits, cooldown;
            cin >> type >> x >> y >> size >> health >> fruits >> cooldown;
            Tree t;
            t.cell = xy(x, y);
            t.type = parseFruit(type);
            t.size = size;
            t.health = health;
            t.fruits = fruits;
            t.cooldown = cooldown;
            trees.push_back(t);
        }

        int trollCount;
        cin >> trollCount;
        vector<Troll> trolls;
        trolls.reserve(trollCount);
        for (int i = 0; i < trollCount; i++) {
            Troll tr;
            int x, y;
            cin >> tr.id >> tr.player >> x >> y
                >> tr.moveSpeed >> tr.carryCap >> tr.harvestPower >> tr.chopPower
                >> tr.carry[ITEM_PLUM] >> tr.carry[ITEM_LEMON]
                >> tr.carry[ITEM_APPLE] >> tr.carry[ITEM_BANANA]
                >> tr.carry[ITEM_IRON] >> tr.carry[ITEM_WOOD];
            tr.cell = xy(x, y);
            trolls.push_back(tr);
        }

        string out;
        for (const auto& tr : trolls) {
            if (tr.player != PLAYER_ME) continue;
            if (!out.empty()) out += ";";
            out += decideTroll(tr, trees);
        }
        if (out.empty()) out = "WAIT";
        cout << out << endl;
    }

    return 0;
}
